From cb0a80111a29b4dbe1af91f499230cb43d4b5707 Mon Sep 17 00:00:00 2001
From: Yann Dirson <yann.dirson@vates.tech>
Date: Mon, 8 Dec 2025 18:39:00 +0100
Subject: [PATCH 6/6] WIP: Q&D port from yum to dnf

APIs have vastly changed.

This is "just enough for check_update calls not to bomb out and show
updates", so we don't have an infinite spinner in xolite for the list
of patches.

* only check_updates was tested
* there was some transaction stuff, but what it was doing is not clear.
  Ordering the packages?  It was just removed for now.  Do we care?
* preconf stuff was ignored:
  * setting debuglevel to 0 may be useful, needs to determine how to do it
  * setting plugins to false possibly a bad idea anyway, we need "priority"
    (but is it still in a plugin?)
* the run_command('yum') was replaced by 'dnf' but untested, it should
  anyway probably be replaced by the suitable library call
---
 SOURCES/etc/xapi.d/plugins/updater.py | 42 +++++++++++++--------------
 1 file changed, 20 insertions(+), 22 deletions(-)

diff --git a/SOURCES/etc/xapi.d/plugins/updater.py b/SOURCES/etc/xapi.d/plugins/updater.py
index 418f9a0..e058dba 100755
--- a/SOURCES/etc/xapi.d/plugins/updater.py
+++ b/SOURCES/etc/xapi.d/plugins/updater.py
@@ -4,7 +4,7 @@
 from functools import wraps
 import configparser
 import json
-import yum
+import dnf
 
 import XenAPIPlugin
 
@@ -97,8 +97,11 @@ def operationlock(*pid_args, **pid_kwargs):
 
 
 def display_package(p):
-    if len(p.changelog):
-        changelog = {'date': p.changelog[0][0], 'author': p.changelog[0][1], 'description': p.changelog[0][2]}
+    if p.changelogs:
+        # only shows last entry, untested
+        changelog = {'date': p.changelogs[0]['timestamp'],
+                     'author': p.changelogs[0]['author'],
+                     'description': p.changelogs[0]['text']}
     else:
         changelog = None
     return {'name': p.name, 'version': p.version, 'release': p.release, 'description': p.summary,
@@ -117,10 +120,9 @@ def install_helper(session, args, action):
     if action == 'install' and not packages:
         raise Exception('Missing or empty argument `packages`')
 
-    yum_instance = yum.YumBase()
-    yum_instance.preconf.debuglevel = 0
-    yum_instance.preconf.plugins = False
-    repos = build_repo_list(yum_instance.repos.listEnabled(), args.get('repos'))
+    yum_instance = dnf.Base()
+    yum_instance.read_all_repos()
+    repos = build_repo_list(list(yum_instance.repos.iter_enabled()), args.get('repos'))
 
     task = None
     res = None
@@ -137,7 +139,7 @@ def install_helper(session, args, action):
                 action.capitalize(), packages, host_name, host_uuid
             ), '')
 
-        command = ['yum', action, '--disablerepo=*', '--enablerepo=' + ','.join(r.id for r in repos), '-y']
+        command = ['dnf', action, '--disablerepo=*', '--enablerepo=' + ','.join(r.id for r in repos), '-y']
         if packages:
             command.append(packages)
         res = run_command(command)
@@ -159,20 +161,16 @@ def install(session, args):
 @error_wrapped
 @operationlock()
 def check_update(session, args):
-    yum_instance = yum.YumBase()
-    yum_instance.preconf.debuglevel = 0
-    yum_instance.preconf.plugins = True
-    repos = build_repo_list(yum_instance.repos.listEnabled(), args.get('repos'))
-    yum_instance.repos.disableRepo('*')
-    yum_instance.repos.enableRepo(','.join(r.id for r in repos))
-    packages = yum_instance.doPackageLists(pkgnarrow='updates')
-    del yum_instance.ts
-    yum_instance.initActionTs()  # make a new, blank ts to populate
-    yum_instance.populateTs(keepold=0)
-    yum_instance.ts.check()  # required for ordering
-    yum_instance.ts.order()
-    # Python 2&3 compatible code
-    return json.dumps(list(map(display_package, packages)))
+    yum_instance = dnf.Base()
+    yum_instance.read_all_repos()
+    repos = build_repo_list(list(yum_instance.repos.iter_enabled()), args.get('repos'))
+    for repo in yum_instance.repos.iter_enabled():
+        repo.disable()
+    for repo in repos:
+        repo.enable()
+    yum_instance.fill_sack()
+    packages = yum_instance.sack.query().available().upgrades()
+    return json.dumps([display_package(p) for p in packages])
 
 @error_wrapped
 @operationlock(timeout=10)
-- 
2.47.3

